api/notifiers.go

```diff
@@ -0,0 +1,217 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+const NotifierTypeEmail = "EmailNotifier"
+const NotifierTypeOpsGenie = "OpsGenieNotifier"
+const NotifierTypePagerDuty = "PagerDutyNotifier"
+const NotifierTypeSlack = "SlackNotifier"
+const NotifierTypeVictorOps = "VictorOpsNotifier"
+const NotifierTypeWebHook = "WebHookNotifier"
+
+type Notifiers struct {
+	client *Client
+}
+
+type Notifier struct {
+	Entity     string                 `json:"entity"`
+	ID         string                 `json:"id,omitempty"`
+	Name       string                 `json:"name"`
+	Properties map[string]interface{} `json:"properties"`
+}
+
+func (c *Client) Notifiers() *Notifiers { return &Notifiers{client: c} }
+
+func (n *Notifiers) List(view string) ([]Notifier, error) {
+	url := fmt.Sprintf("api/v1/repositories/%s/alertnotifiers", view)
+
+	res, err := n.client.HttpGET(url)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	notifiers := []Notifier{}
+	jsonErr := json.Unmarshal(body, &notifiers)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+
+	return notifiers, nil
+}
+
+func (n *Notifiers) Update(viewName string, notifier *Notifier) (*Notifier, error) {
+	existingID, err := n.convertNotifierNameToID(viewName, notifier.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert notifier name to id: %v", err)
+	}
+	url := fmt.Sprintf("api/v1/repositories/%s/alertnotifiers/%s", viewName, existingID)
+
+	jsonStr, err := json.Marshal(notifier)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert notifier to json string: %v", err)
+	}
+
+	res, err := n.client.HttpPUT(url, bytes.NewBuffer(jsonStr))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	fmt.Println(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	resNotifier := Notifier{}
+	jsonErr := json.Unmarshal(body, &resNotifier)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+
+	return &resNotifier, nil
+}
+
+func (n *Notifiers) Add(viewName string, notifier *Notifier, force bool) (*Notifier, error) {
```

Comment:
Same as the above with the `Add()` and `Update()` methods (duplication). As well as the duplication around the unmarshalling in the other public methods here.

Reply:
After pulling out the marshal/unmarshal code, I'm not sure it would help that much to move this out as well. I'm thinking of just leaving this as-is unless you think I should still do it.